Sources/KlaviyoSwiftExtension/KlaviyoExtension.swift

```diff
@@ -0,0 +1,139 @@
+            contentHandler(bestAttemptContent)
+            return
+        }
+
+        // 1b.falling back to .png in case the media type isn't sent from the server.
+        let imageTypeString = bestAttemptContent.userInfo["rich-media-type"] as? String ?? "png"
+
+        // 2. once we have the url lets download the media from the server
+        downloadMedia(for: imageURLString) { localFileURL in
+            guard let localFileURL = localFileURL else {
+                contentHandler(bestAttemptContent)
+                return
+            }
+
+            let localFilePathWithTypeString = "\(localFileURL.path).\(imageTypeString)"
+
+            // 3. once we have the local file URL we will create an attachment
+            createAttachment(
+                localFileURL: localFileURL,
+                localFilePathWithTypeString: localFilePathWithTypeString) { attachment in
+                    guard let attachment = attachment else {
+                        contentHandler(bestAttemptContent)
+                        return
+                    }
+
+                    // 4. assign the create attachement to the best attempt content attachment and call the content handler so that the notification with the
+                    //    media can be delivered to the user.
+                    bestAttemptContent.attachments = [attachment]
+                    contentHandler(bestAttemptContent)
+                }
+        }
+    }
+    
+    public func handleNotificationServiceExtensionTimeWillExpireRequest(
+        request: UNNotificationRequest,
+        bestAttemptContent: UNMutableNotificationContent,
+        contentHandler: @escaping (UNNotificationContent) -> Void
+    ) {
+        contentHandler(bestAttemptContent)
+    }
+    
+    /// downloads the media from the provided URL and writes to disk and provides a URL to the data on disk
+    /// - Parameters:
+    ///   - urlString: the URL from where the media needs to be downloaded
+    ///   - completion: closure that would be called when the image has finished downloading and the URL to the data on disk is available.
+    ///                 note that in the case of failure the closure will still be called but with `nil`.
+    private func downloadMedia(
+        for urlString: String,
+        completion: @escaping (URL?) -> Void) {
+        guard let imageURL = URL(string: urlString) else {
+            completion(nil)
+            return
+        }
+
+        let task = URLSession.shared.downloadTask(with: imageURL) { file, _, error in
+            if let error = error {
+                print("error when downloading push media = \(error.localizedDescription)")
+                completion(nil)
+                return
+            }
+
+            guard let file = file else {
+                completion(nil)
+                return
+            }
+
+            completion(file)
+        }
+        task.resume()
```

Comment:
We don't need to hold onto the task right?

Reply:
Was also thinking should we try to use async api where available?

Reply:
Yeah we don't. I removed it. With using async await - I'll probably have it as a fast follow since we've already tested this code now.